src/systems/implicit_system.C

```diff
@@ -675,21 +673,25 @@ void ImplicitSystem::assemble_residual_derivatives(const ParameterVector& parame
       // (R(p-dp) - R(p+dp)) / (2*dp)
 
       Number old_parameter = *parameters[p];
-      *parameters[p] -= deltap;
+
+      const Real delta_p =
+        TOLERANCE * std::max(std::abs(old_parameter), 1e-3);
```

Comment:
Yeah, it's an ugly hack.  If someone has a parameter whose default value is 0, in the extreme case, the best we can do is "hope TOLERANCE*1e-3 works fine", because the alternative is "divide-by-zero error".

"Let users override delta_p on a per-parameter basis" would be better, and "Let users supply analytic derivatives" better still, but this works for us for now.